src/_blog/libp2p-webrtc-browser-to-server.md

```diff
@@ -0,0 +1,223 @@
+
+**Table of Contents**
+
+_IN DRAFT, NOT READY FOR REVIEW_
+
+[[toc]]
+
+## Overview
+
+The [libp2p project](https://libp2p.io) specifies many [transports](https://libp2p.io/implementations/#transports) developed in a variety of implementations (Rust, Go,  and JavaScript).  While connecting servers to servers has been around for a while via TCP and QUIC, there haven't been optimal solutions for connecting browsers to servers that work on a broad spectrum of browsers.  Additionally, there was a need to establish browser-to-browser connectivity in a decentralized way.
+
+Other transports laid the groundwork for a WebRTC implementation in libp2p, but had some shortcomings, including centralized signaling and the need for a trusted TLS certificate, that created a need for a formal specification. [Little Bear Labs](https://littlebearlabs.io/) teamed up with Protocol Labs and the libp2p community to define a specification and work on the SDK implementation.  Protocol Labs authored the [Rust](https://github.com/libp2p/rust-libp2p) implementation, while Little Bear Labs focused on the [Go](https://github.com/libp2p/go-libp2p) and [JavaScript](https://github.com/libp2p/js-libp2p-webrtc) implementations.
+
+Before we deep dive into the WebRTC implementation within libp2p, let's first understand what WebRTC is and how it's currently being used.
+
+## WebRTC in the Browser
+
+`Web Real-Time Communication`, commonly referred to as `WebRTC`, is a [set of standards](https://w3c.github.io/webrtc-pc/) that allows browsers, as well as other clients and servers, to connect to other peers in order to exchange audio, video, and data.  In most cases, peers are directly connected to other peers, allowing for a more private experience and fewer hops than on a relay.
+
+While WebRTC handles audio, video, and data traffic, we're just going to focus on the data aspect because that's the API leveraged in libp2p-webrtc.
+
+WebRTC is built directlry into browsers, so using the API is a simple and straightfoward task.  Peers connect to each other via a `RTCPeerConnection` interface.  Once connected, `RTCDataChannels` can be added to the connection to send and receive binary data.
+
+![](https://i.imgur.com/Zv221BT.png)
+
+<!--
+participantspacing 7
+entryspacing 0.6
+
+Peer A->STUN: Who Am I?
+STUN->Peer A: Symmetric Nat
+Peer A->TURN: Channel Please
+Peer A->Signaling Channel: Offer SDP
+Signaling Channel->Peer B: Offer SDP
+Peer B->STUN: Who Am I?
+STUN->Peer B: 159.225.242.189
+Peer B->Signaling Channel: Answer SDP
+Signaling Channel->Peer A: Answer SDP
+
+
+Peer A->(1)Signaling Channel: DTLS Handshake
+Signaling Channel->(1)Peer B: DTLS Handshake
+Peer B->(1)Signaling Channel: DTLS Handshake
+Signaling Channel->(1)Peer A: DTLS Handshake
+
+
+Peer A<->Signaling Channel:Duplex Send/Receive SCTP data, encrypted with DTLS
+Signaling Channel<->Peer B:Duplex Send/Receive SCTP data, encrypted with DTLS
+-->
```

Comment:
I assume you want to use this super complicated diagram to show how we simplified things in libp2p. I find it more confusing than helpful though. For example, messages seem to pass through the STUN and the TURN server (which they don’t for the STUN server, and only sometimes do for TURN).

TURN is not really relevant in this context anyway, since we’re explicitly dealing with the browser to server use case here.

Reply:
That's a good point @marten-seemann.  The original intent was to give a quick deep-dive into WebRTC, but I can see that it takes away from the overall point of the post.  I'm going to remove the diagram.